src/node/witness.rs

```diff
@@ -151,21 +151,49 @@ impl<J: Jet> WitnessNode<J> {
         .expect("type inference won't fail if it succeeded before")
     }
 
-    pub fn finalize(&self) -> Result<Arc<RedeemNode<J>>, Error> {
-        // 0. Setup some structure for the WitnessNode->RedeemNode conversion
+    /// Finalize the witness program as an unpruned redeem program.
+    ///
+    /// Witness nodes must be populated with sufficient data,
+    /// to ensure that the resulting redeem program successfully runs on the Bit Machine.
+    /// Furthermore, **all** disconnected branches must be populated,
+    /// even those that are not executed.
+    ///
+    /// The resulting redeem program is **not pruned**.
+    ///
+    /// ## See
+    ///
+    /// [`RedeemNode::prune`]
+    pub fn finalize_unpruned(&self) -> Result<Arc<RedeemNode<J>>, Error> {
         struct Finalizer<J>(PhantomData<J>);
 
         impl<J: Jet> Converter<Witness<J>, Redeem<J>> for Finalizer<J> {
             type Error = Error;
+
             fn convert_witness(
                 &mut self,
-                _: &PostOrderIterItem<&WitnessNode<J>>,
+                data: &PostOrderIterItem<&WitnessNode<J>>,
                 wit: &Option<Value>,
             ) -> Result<Value, Self::Error> {
                 if let Some(ref wit) = wit {
                     Ok(wit.shallow_clone())
                 } else {
-                    Err(Error::IncompleteFinalization)
+                    // We insert a zero value into unpopulated witness nodes,
```

Comment:
In 40e5ab946fd5dfaa7fd97bd936d3918a3a7919bc:

I guess, if we had hidden nodes in our `Node` enum, we could handle this more cleanly (and also get rid of the "you must populate every disconnect node" requirement below)?